File: src/main/java/com/redeprivada/sorting/HeapSort.java
package com.redeprivada.sorting;

public class HeapSort implements Sortable {

    @Override
    public void sort(int[] nums) {
        int heapsize = nums.length;
        buildmaxheap(nums, heapsize);
        for (int i = heapsize - 1; i >= 1; --i) {
            int temp = nums[0];
            nums[0] = nums[i];
            nums[i] = temp;
            heapsize--;
            maxheapify(nums, heapsize, 0);
        }
    }

    private void maxheapify(int[] nums, int heapsize, int i) {
        int l = (i + 1) * 2 - 1;
        int r = (i + 1) * 2;
        int largest = i;
        if (l < heapsize && nums[l] > nums[i])
            largest = l;
        if (r < heapsize && nums[r] > nums[largest])
            largest = r;
        if (largest != i) {
            int temp = nums[i];
            nums[i] = nums[largest];
            nums[largest] = temp;
            maxheapify(nums, heapsize, largest);
        }
    }

    private void buildmaxheap(int[] nums, int heapsize) {
        for (int i = heapsize / 2; i >= 0; --i) {
            maxheapify(nums, heapsize, i);
        }
    }
}

File: src/main/kotlin/com/redeprivada/sorting/HeapSort.kt
package com.redeprivada.sorting

class HeapSort : Sortable {
    override fun sort(nums: IntArray) {
        var heapsize = nums.size
        buildmaxheap(nums, heapsize)
        for (i in heapsize - 1 downTo 1) {
            val temp = nums[0]
            nums[0] = nums[i]
            nums[i] = temp
            heapsize--
            maxheapify(nums, heapsize, 0)
        }
    }

    private fun maxheapify(nums: IntArray, heapsize: Int, i: Int) {
        val l = (i + 1) * 2 - 1
        val r = (i + 1) * 2
        var largest = i
        if (l < heapsize && nums[l] > nums[i]) largest = l
        if (r < heapsize && nums[r] > nums[largest]) largest = r
        if (largest != i) {
            val temp = nums[i]
            nums[i] = nums[largest]
            nums[largest] = temp
            maxheapify(nums, heapsize, largest)
        }
    }

    private fun buildmaxheap(nums: IntArray, heapsize: Int) {
        for (i in heapsize / 2 downTo 0) {
            maxheapify(nums, heapsize, i)
        }
    }
}
